src/base.rs

```diff
@@ -0,0 +1,465 @@
+            fd
+        } else {
+            // Timer
+            -1
+        };
+
+        let callback_ctx = if let Some(ctx) = callback_ctx {
+            ctx
+        } else {
+            std::ptr::null::<c_void>() as *mut std::ffi::c_void
+        };
+
+        unsafe {
+            libevent_sys::event_assign(
+                ev.as_ptr(),
+                self.as_raw().as_ptr(),
+                fd,
+                flags.bits() as c_short,
+                Some(callback),
+                callback_ctx,
+            )
+        }
+    }
+
+    /// Wrapper for libevent's `event_add`, which activates an initialized
+    /// `event` for a pre-defined `Base` and a given timeout interval.
+    pub fn event_add(
+        &self,
+        event: NonNull<libevent_sys::event>,
+        timeout: Option<Duration>,
+    ) -> c_int {
+        unsafe {
+            let p = event.as_ptr();
+            if let Some(tv) = timeout {
+                libevent_sys::event_add(p, &to_timeval(tv))
+            } else {
+                // null timeout means no timeout to libevent
+                libevent_sys::event_add(p, std::ptr::null())
+            }
+        }
+    }
+}
+
+unsafe impl Send for Base {}
+
+impl<S, T: Exec<S, F>, F> EventCallbackWrapper<S, T, F> {
+    pub fn new(inner: F, event: Event<S>) -> Box<Self> {
+        Box::new(Self {
+            inner,
+            event: Some(event),
+            _phantom: std::marker::PhantomData::default(),
+        })
+    }
+}
+
+/// Handles freeing the callback wrapper memory.
+///
+/// It is set up in a fashion similar to `handle_wrapped_callback` so that
+/// callers do not need to know the type parameters, only that they have a
+/// function pointer to a monomorphized function that is valid for a given
+/// event context.
+pub(crate) unsafe extern "C" fn finalize_wrapped_callback<S, T, F>(
+    event: *mut libevent_sys::event,
+    ctx: EventCallbackCtx,
+) {
+    // Wrapper was allocated with Box, now free it with Drop.
+    let cb: *mut EventCallbackWrapper<S, T, F> = ctx as *mut EventCallbackWrapper<S, T, F>;
+    let owned_cb = Box::from_raw(cb);
+    drop(owned_cb);
+
+    // Now clear the event's ctx pointer field.
+    let null_ctx = std::ptr::null::<c_void>() as *mut std::ffi::c_void;
+    (*event).ev_evcallback.evcb_arg = null_ctx;
+}
+
+/// Acts as a C-compatible trampoline for the user-defined callback closure.
+pub(crate) extern "C" fn handle_wrapped_callback<S, T, F>(
+    fd: EvutilSocket,
+    event: std::os::raw::c_short,
+    ctx: EventCallbackCtx,
+) where
+    T: Exec<S, F>,
+{
+    let cb_ref = unsafe {
+        let cb: *mut EventCallbackWrapper<S, T, F> = ctx as *mut EventCallbackWrapper<S, T, F>;
+        let _cb_ref: &mut EventCallbackWrapper<S, T, F> = &mut *cb;
+        _cb_ref
+    };
+
+    let flags = EventFlags::from_bits_truncate(event as u32);
+    let ev = cb_ref.event.as_mut().expect("Missing event for callback");
+
+    ev.set_in_callback(true);
```

Comment:
Does the setting of `in_callback` serve a purpose at present?  I'm not against it, but I'm just wondering if I'm not seeing where it is being used currently.